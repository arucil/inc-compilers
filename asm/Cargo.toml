[package]
name = "asm"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
indexmap = "1.8.1"
enum-iterator = "0.8.1"
num-derive = "0.3.3"
num-traits = "0.2.15"
support = { path = "../support" }
