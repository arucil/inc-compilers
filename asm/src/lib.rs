#![feature(never_type)]

use num_derive::{FromPrimitive, ToPrimitive};
use std::fmt::{self, Debug, Formatter, Write};

#[derive(Debug, Clone)]
pub struct Program<INFO = (), VAR = !> {
  pub info: INFO,
  pub blocks: Vec<(String, Block<VAR>)>,
}

#[derive(Clone)]
pub struct Block<VAR = !> {
  pub code: Vec<Instr<VAR>>,
}

#[derive(Clone)]
#[non_exhaustive]
pub enum Instr<VAR = !> {
  Add(Arg<VAR>, Arg<VAR>),
  Sub(Arg<VAR>, Arg<VAR>),
  Mov(Arg<VAR>, Arg<VAR>),
  Neg(Arg<VAR>),
  Call(String, usize),
  Ret,
  Push(Arg<VAR>),
  Pop(Arg<VAR>),
  Jmp(String),
  Syscall,
}

#[derive(Clone, PartialEq, Eq)]
pub enum Arg<VAR = !> {
  Imm(i64),
  Reg(Reg),
  Deref(Reg, i32),
  Var(VAR),
}

#[derive(
  Clone, Copy, FromPrimitive, ToPrimitive, PartialEq, Eq, PartialOrd, Ord, Hash,
)]
pub enum Reg {
  Rsp,
  Rbp,
  Rax,
  Rbx,
  Rcx,
  Rdx,
  Rsi,
  Rdi,
  R8,
  R9,
  R10,
  R11,
  R12,
  R13,
  R14,
  R15,
}

impl<INFO: Debug, VAR: Debug> Program<INFO, VAR> {
  pub fn to_string_pretty(&self) -> String {
    let mut buf = format!("{:?}", self.info);
    for (label, block) in &self.blocks {
      writeln!(&mut buf, "{}:", label).unwrap();
      buf += &format!("{:?}", block);
    }
    buf
  }

  pub fn to_nasm(&self) -> String {
    let mut buf = format!(
      "extern read_int, print_int, print_newline\n\
      section .text\n"
    );
    for (label, block) in &self.blocks {
      buf += "\n";
      if label == "_start" {
        buf += "    global _start\n";
      }
      writeln!(&mut buf, "{}:", label).unwrap();
      buf += &format!("{:?}", block);
    }
    buf
  }
}

impl<VAR: Debug> Debug for Block<VAR> {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    for instr in &self.code {
      write!(f, "    ")?;
      instr.fmt(f)?;
      writeln!(f)?;
    }
    Ok(())
  }
}

impl<VAR: Debug> Debug for Instr<VAR> {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    match self {
      Self::Add(src, dest) => write!(f, "add {:?}, {:?}", dest, src),
      Self::Mov(src, dest) => write!(f, "mov {:?}, {:?}", dest, src),
      Self::Call(label, _) => write!(f, "call {}", label),
      Self::Jmp(label) => write!(f, "jmp {}", label),
      Self::Neg(dest) => write!(f, "neg {:?}", dest),
      Self::Pop(dest) => write!(f, "pop {:?}", dest),
      Self::Push(src) => write!(f, "push {:?}", src),
      Self::Ret => write!(f, "ret"),
      Self::Sub(src, dest) => write!(f, "sub {:?}, {:?}", dest, src),
      Self::Syscall => write!(f, "syscall"),
    }
  }
}

impl<VAR: Debug> Debug for Arg<VAR> {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    match self {
      Self::Imm(n) => write!(f, "{}", n),
      Self::Var(var) => var.fmt(f),
      Self::Deref(r, i) => {
        if *i > 0 {
          write!(f, "qword [{:?} + {}]", r, i)
        } else if *i == 0 {
          write!(f, "qword [{:?}]", r)
        } else {
          write!(f, "qword [{:?} - {}]", r, -i)
        }
      }
      Self::Reg(r) => r.fmt(f),
    }
  }
}

impl Reg {
  pub fn is_callee_saved(&self) -> bool {
    match self {
      Reg::Rsp
      | Reg::Rbp
      | Reg::Rbx
      | Reg::R12
      | Reg::R13
      | Reg::R14
      | Reg::R15 => true,
      _ => false,
    }
  }
}

impl Debug for Reg {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    use Reg::*;
    let op = match self {
      Rsp => "rsp",
      Rbp => "rbp",
      Rax => "rax",
      Rbx => "rbx",
      Rcx => "rcx",
      Rdx => "rdx",
      Rsi => "rsi",
      Rdi => "rdi",
      R8 => "r8",
      R9 => "r9",
      R10 => "r10",
      R11 => "r11",
      R12 => "r12",
      R13 => "r13",
      R14 => "r14",
      R15 => "r15",
    };
    f.write_str(op)
  }
}
