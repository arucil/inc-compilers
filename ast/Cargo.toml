[package]
name = "ast"
version = "0.1.0"
edition = "2018"

[dependencies]
pretty = "0.10.0"
support = { path = "../support" }

[dev-dependencies]
pretty_assertions = "0.7.2"
