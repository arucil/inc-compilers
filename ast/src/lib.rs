#![feature(never_type, box_syntax)]

use id_arena::{Arena, Id};
use indexmap::IndexMap;
use pretty::*;
use std::fmt::{self, Debug, Display, Formatter};
use std::iter;
use support::Range;

pub mod parser;

pub use parser::{parse, Result};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Program<VAR = String, TYPE = ()> {
  pub type_defs: IndexMap<String, Type>,
  pub fun_defs: IndexMap<String, FunDef<VAR, TYPE>>,
  pub body: Exp<VAR, TYPE>,
  pub types: Arena<Type>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FunDef<VAR = String, TYPE = ()> {
  pub params: Vec<(VAR, Type)>,
  pub ret: Type,
  pub body: Exp<VAR, TYPE>,
  pub range: Range,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Exp<VAR = String, TYPE = ()> {
  pub kind: ExpKind<VAR, TYPE>,
  pub range: Range,
  pub ty: TYPE,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ExpKind<VAR, TYPE> {
  Int(i64),
  Prim {
    op: (Range, &'static str),
    args: Vec<Exp<VAR, TYPE>>,
  },
  Apply {
    fun: Box<Exp<VAR, TYPE>>,
    args: Vec<Exp<VAR, TYPE>>,
    r#struct: Option<StructApp>,
  },
  Var(VAR),
  Str(String),
  Let {
    var: (Range, VAR),
    init: Box<Exp<VAR, TYPE>>,
    body: Box<Exp<VAR, TYPE>>,
  },
  Bool(bool),
  If {
    cond: Box<Exp<VAR, TYPE>>,
    conseq: Box<Exp<VAR, TYPE>>,
    alt: Box<Exp<VAR, TYPE>>,
  },
  Set {
    var: (Range, VAR),
    exp: Box<Exp<VAR, TYPE>>,
  },
  Get(VAR),
  FunRef {
    name: String,
    arity: usize,
  },
  Begin {
    seq: Vec<Exp<VAR, TYPE>>,
    last: Box<Exp<VAR, TYPE>>,
  },
  While {
    cond: Box<Exp<VAR, TYPE>>,
    body: Box<Exp<VAR, TYPE>>,
  },
  Void,
  Print(
    /// nonempty
    Vec<Exp<VAR, TYPE>>,
  ),
  NewLine,
  Error(Error<VAR, TYPE>),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum StructApp {
  Ctor,
  Getter(u32),
  Setter(u32),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Error<VAR, TYPE> {
  Length(Box<Exp<VAR, TYPE>>),
  OutOfBounds {
    index: Box<Exp<VAR, TYPE>>,
    len: Box<Exp<VAR, TYPE>>,
  },
  DivByZero,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Type {
  Void,
  Int,
  Bool,
  Str,
  Tuple(Vec<Type>),
  Array(Box<Type>),
  Alias(TypeAlias),
  Struct(IndexMap<String, Type>),
  Fun { params: Vec<Type>, ret: Box<Type> },
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TypeAlias {
  Pending(Range, String),
  Resolved(TypeId),
}

pub type TypeId = Id<Type>;

#[derive(Clone, PartialEq, Eq, Hash)]
pub struct IdxVar {
  pub name: String,
  pub index: usize,
}

impl IdxVar {
  pub fn new<S: ToString>(name: S) -> Self {
    let name = name.to_string();
    if let Some(i) = name.find('.') {
      if let Ok(index) = name[i + 1..].parse::<usize>() {
        return Self {
          name: name[..i].to_owned(),
          index,
        };
      }
    }
    Self { name, index: 0 }
  }
}

impl Display for IdxVar {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    write!(f, "{}.{}", self.name, self.index)
  }
}

impl Debug for IdxVar {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    Display::fmt(self, f)
  }
}

impl<VAR: Display, TYPE: Debug> Program<VAR, TYPE> {
  pub fn to_string_pretty(&self) -> String {
    let doc = self.to_doc();
    let mut buf = String::new();
    doc.render_fmt(50, &mut buf).unwrap();
    buf
  }

  fn to_doc(&self) -> RcDoc {
    RcDoc::intersperse(
      self
        .fun_defs
        .iter()
        .map(|(name, def)| def.to_doc(name))
        .chain(std::iter::once(self.body.to_doc())),
      Doc::line(),
    )
  }
}

impl<VAR: Display, TYPE: Debug> FunDef<VAR, TYPE> {
  fn to_doc<'a>(&'a self, name: &str) -> RcDoc<'a> {
    RcDoc::text("(")
      .append(
        RcDoc::text("define")
          .append(Doc::line())
          .append(
            RcDoc::text("(")
              .append(
                RcDoc::text(name.to_owned())
                  .append(Doc::line())
                  .append(RcDoc::intersperse(
                    self.params.iter().map(|(name, ty)| {
                      RcDoc::text("[")
                        .append(name.to_string())
                        .append(Doc::line())
                        .append(ty.to_doc())
                        .append("]")
                        .group()
                    }),
                    Doc::line(),
                  ))
                  .nest(1),
              )
              .append(")")
              .group(),
          )
          .group()
          .append(Doc::space())
          .append(":")
          .append(Doc::space())
          .append(self.ret.to_doc())
          .group()
          .append(Doc::line())
          .append(self.body.to_doc())
          .nest(1),
      )
      .append(")")
  }
}

impl<VAR: Display, TYPE: Debug> Exp<VAR, TYPE> {
  fn to_doc(&self) -> RcDoc {
    match &self.kind {
      ExpKind::Int(n) => RcDoc::text(format!("{}", n)),
      ExpKind::Str(s) => RcDoc::text(format!("{:?}", s)),
      ExpKind::Var(var) => RcDoc::text(format!("{}", var)),
      ExpKind::Get(var) => {
        RcDoc::text("{").append(format!("{}", var)).append("}")
      }
      ExpKind::Prim { op, args } => RcDoc::text("(")
        .append(
          RcDoc::intersperse(
            iter::once(RcDoc::text(op.1))
              .chain(args.iter().map(|arg| arg.to_doc())),
            Doc::line(),
          )
          .nest(1)
          .group(),
        )
        .append(")"),
      ExpKind::Apply {
        fun,
        args,
        r#struct: _,
      } => RcDoc::text("(")
        .append(
          RcDoc::intersperse(
            iter::once(fun.to_doc()).chain(args.iter().map(|arg| arg.to_doc())),
            Doc::line(),
          )
          .nest(1)
          .group(),
        )
        .append(")"),
      ExpKind::Let { var, init, body } => RcDoc::text("(")
        .append(
          RcDoc::text("let")
            .append(Doc::line())
            .append(
              RcDoc::text("[")
                .append(
                  RcDoc::text(format!("{}", var.1))
                    .append(Doc::line())
                    .append(init.to_doc())
                    .nest(1)
                    .group(),
                )
                .append("]"),
            )
            .append(Doc::line())
            .append(body.to_doc())
            .nest(1),
        )
        .append(")"),
      ExpKind::Bool(true) => RcDoc::text("#t"),
      ExpKind::Bool(false) => RcDoc::text("#f"),
      ExpKind::If { cond, conseq, alt } => RcDoc::text("(")
        .append(
          RcDoc::text("if")
            .append(Doc::line())
            .append(cond.to_doc())
            .group()
            .append(Doc::line())
            .append(conseq.to_doc())
            .append(Doc::line())
            .append(alt.to_doc())
            .nest(1),
        )
        .append(")"),
      ExpKind::Set { var, exp } => RcDoc::text("(")
        .append(
          RcDoc::text("set!")
            .append(Doc::line())
            .append(format!("{}", var.1))
            .append(Doc::line())
            .append(exp.to_doc())
            .nest(1)
            .group(),
        )
        .append(")"),
      ExpKind::Begin { seq, last } => RcDoc::text("(")
        .append(
          RcDoc::text("begin")
            .append(Doc::line())
            .append(RcDoc::intersperse(
              seq
                .iter()
                .chain(std::iter::once(&**last))
                .map(|exp| exp.to_doc()),
              Doc::line(),
            ))
            .nest(1),
        )
        .append(")"),
      ExpKind::While { cond, body } => RcDoc::text("(")
        .append(
          RcDoc::text("while")
            .append(Doc::line())
            .append(cond.to_doc())
            .group()
            .append(Doc::line())
            .append(body.to_doc())
            .nest(1)
            .group(),
        )
        .append(")"),
      ExpKind::Void => RcDoc::text("(void)"),
      ExpKind::Print(args) => RcDoc::text("(")
        .append(
          RcDoc::intersperse(
            iter::once(RcDoc::text("print"))
              .chain(args.iter().map(|arg| arg.to_doc())),
            Doc::line(),
          )
          .nest(1)
          .group(),
        )
        .append(")"),
      ExpKind::NewLine => RcDoc::text("(newline)"),
      ExpKind::Error(Error::Length(arg)) => RcDoc::text("(")
        .append(
          RcDoc::text("length-error")
            .append(Doc::line())
            .append(arg.to_doc())
            .nest(1)
            .group(),
        )
        .append(")"),
      ExpKind::Error(Error::OutOfBounds { index, len }) => RcDoc::text("(")
        .append(
          RcDoc::text("out-of-bounds-error")
            .append(Doc::line())
            .append(index.to_doc())
            .append(Doc::line())
            .append(len.to_doc())
            .nest(1)
            .group(),
        )
        .append(")"),
      ExpKind::Error(Error::DivByZero) => RcDoc::text("(div-by-zero)"),
      ExpKind::FunRef { name, arity } => RcDoc::text("{")
        .append(
          RcDoc::text("fun-ref")
            .append(Doc::line())
            .append(name)
            .append(Doc::line())
            .append(format!("{}", arity))
            .nest(1)
            .group(),
        )
        .append("}"),
    }
  }
}

impl<VAR, TYPE> ExpKind<VAR, TYPE> {
  pub fn is_atomic(&self) -> bool {
    match self {
      Self::Int(_) => true,
      Self::Prim { .. } => false,
      Self::Apply { .. } => false,
      Self::Var(_) => true,
      Self::Str(_) => true,
      Self::Let { .. } => false,
      Self::Bool(_) => true,
      Self::If { .. } => false,
      Self::Set { .. } => false,
      Self::Get(_) => false,
      Self::FunRef { .. } => true,
      Self::Begin { .. } => false,
      Self::While { .. } => false,
      Self::Void => true,
      Self::Print(..) => false,
      Self::NewLine => false,
      Self::Error(_) => false,
    }
  }
}

impl Type {
  pub fn is_ref(&self, types: &Arena<Type>) -> bool {
    match self {
      Self::Tuple(_) => true,
      Self::Array(_) => true,
      Self::Str => true,
      Self::Alias(TypeAlias::Resolved(id)) => types[*id].is_ref(types),
      Self::Alias(TypeAlias::Pending(..)) => unreachable!(),
      _ => false,
    }
  }

  pub fn resolved(&self, types: &Arena<Type>) -> Self {
    match self {
      Self::Alias(TypeAlias::Resolved(id)) => types[*id].resolved(types),
      Self::Alias(TypeAlias::Pending(..)) => unreachable!(),
      _ => self.clone(),
    }
  }

  fn to_doc(&self) -> RcDoc {
    match self {
      Self::Void => RcDoc::text("Void"),
      Self::Int => RcDoc::text("Int"),
      Self::Bool => RcDoc::text("Bool"),
      Self::Str => RcDoc::text("Str"),
      Self::Tuple(tys) => RcDoc::text("(")
        .append(RcDoc::intersperse(
          tys.iter().map(|ty| ty.to_doc()),
          Doc::line(),
        ))
        .nest(1)
        .append(")"),
      Self::Array(ty) => RcDoc::text("(")
        .append(
          RcDoc::text("array-of")
            .append(Doc::line())
            .append(ty.to_doc())
            .nest(1),
        )
        .append(")"),
      Self::Alias(TypeAlias::Pending(_, alias)) => RcDoc::text(alias),
      Self::Alias(TypeAlias::Resolved(id)) => {
        RcDoc::text(format!("(alias {})", id.index()))
      }
      Self::Struct(fields) => RcDoc::text("(")
        .append(
          RcDoc::text("struct")
            .append(Doc::line())
            .append(RcDoc::intersperse(
              fields.iter().map(|(name, ty)| {
                RcDoc::text("[")
                  .append(name)
                  .append(Doc::line())
                  .append(ty.to_doc())
                  .append("]")
              }),
              Doc::line(),
            ))
            .nest(1),
        )
        .append(")"),
      Self::Fun { params, ret } => RcDoc::text("(")
        .append(RcDoc::intersperse(
          params.iter().map(|ty| ty.to_doc()),
          Doc::line(),
        ))
        .append(Doc::line())
        .group()
        .append("->")
        .append(Doc::line())
        .append(ret.to_doc())
        .nest(1)
        .append(")"),
    }
  }
}
