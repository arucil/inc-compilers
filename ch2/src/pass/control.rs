use ast::{Program, Exp, IdxVar};
use std::fmt::{self, Write};
use indexmap::IndexSet;
use support::{WritePretty, Range};

pub struct CProgram {
  pub info: CInfo,
  pub body: Vec<(String, CTail)>,
}

pub struct CInfo {
  pub locals: IndexSet<IdxVar>,
}

pub enum CTail {
  Return(CExp),
  Seq(CStmt, Box<CTail>)
}

pub enum CStmt {
  Assign {
    var: IdxVar,
    exp: CExp,
  }
}

pub enum CExp {
  Atom(CAtom),
  Prim(CPrim),
}

pub enum CPrim {
  Read,
  Neg(CAtom),
  Add(CAtom, CAtom),
}

pub enum CAtom {
  Int(i64),
  Var(IdxVar)
}

impl CProgram {
  pub fn to_string_pretty(&self) -> String {
    let mut buf = String::new();
    self.info.write(&mut buf).unwrap();
    for (label, tail) in &self.body {
      writeln!(&mut buf, "{}:", label).unwrap();
      tail.write(&mut buf).unwrap();
    }
    buf
  }
}

impl WritePretty for CInfo {
  fn write(&self, f: &mut impl Write) -> fmt::Result {
    writeln!(f, "locals: {:?}\n", self.locals)
  }
}

impl WritePretty for CTail {
  fn write(&self, f: &mut impl Write) -> fmt::Result {
    match self {
      Self::Return(exp) => {
        write!(f, "    return ")?;
        exp.write(f)
      }
      Self::Seq(stmt, tail) => {
        write!(f, "    ")?;
        stmt.write(f)?;
        write!(f, "\n")?;
        tail.write(f)
      }
    }
  }
}

impl WritePretty for CStmt {
  fn write(&self, f: &mut impl Write) -> fmt::Result {
    match self {
      Self::Assign { var, exp } => {
        write!(f, "{:?} = ", var)?;
        exp.write(f)
      }
    }
  }
}

impl WritePretty for CExp {
  fn write(&self, f: &mut impl Write) -> fmt::Result {
    match self {
      Self::Atom(atom) => atom.write(f),
      Self::Prim(prim) => prim.write(f),
    }
  }
}

impl WritePretty for CPrim {
  fn write(&self, f: &mut impl Write) -> fmt::Result {
    match self {
      Self::Add(arg1, arg2) => {
        write!(f, "(+ ")?;
        arg1.write(f)?;
        write!(f, " ")?;
        arg2.write(f)?;
        write!(f, ")")
      }
      Self::Neg(arg) => {
        write!(f, "(- ")?;
        arg.write(f)?;
        write!(f, ")")
      }
      Self::Read => {
        write!(f, "(read)")
      }
    }
  }
}

impl WritePretty for CAtom {
  fn write(&self, f: &mut impl Write) -> fmt::Result {
    match self {
      Self::Int(n) => write!(f, "{}", n),
      Self::Var(n) => write!(f, "{:?}", n),
    }
  }
}

pub fn explicate_control(mut prog: Program<IdxVar>) -> CProgram {
  let locals = collect_locals(&prog);
  CProgram {
    info: CInfo {
      locals,
    },
    body: vec![
      ("start".to_owned(), explicate_tail(prog.body.pop().unwrap().1))
    ]
  }
}

fn collect_locals(
  prog: &Program<IdxVar>,
) -> IndexSet<IdxVar> {
  let mut locals = IndexSet::new();
  for (_, exp) in &prog.body {
    collect_exp_locals(exp, &mut locals);
  }
  locals
}

fn collect_exp_locals(
  exp: &Exp<IdxVar>,
  locals: &mut IndexSet<IdxVar>,
) {
  match exp {
    Exp::Int(_) => {}
    Exp::Var(var) => {
      locals.insert(var.clone());
    }
    Exp::Str(_) => {}
    Exp::Let { var, init, body } => {
      locals.insert(var.1.clone());
      collect_exp_locals(&init.1, locals);
      collect_exp_locals(&body.1, locals);
    }
    Exp::Prim { op: _, args } => {
      for (_, arg) in args {
        collect_exp_locals(arg, locals);
      }
    }
    _ => unimplemented!()
  }
}

fn explicate_tail(exp: Exp<IdxVar>) -> CTail {
  match exp {
    exp@(Exp::Int(_) | Exp::Var(_)) => {
      CTail::Return(CExp::Atom(atom(exp)))
    }
    ast::Exp::Prim { op: (_, op), args } => {
      CTail::Return(CExp::Prim(prim(op, args)))
    }
    ast::Exp::Let { var, init, body } => {
      explicate_assign(var.1, init.1, explicate_tail(body.1))
    }
    exp => unimplemented!("unsupported form {:?}", exp)
  }
}

fn explicate_assign(
  var: IdxVar,
  init: ast::Exp<IdxVar>,
  cont: CTail,
) -> CTail {
  match init {
    exp@(Exp::Int(_) | Exp::Var(_)) => {
      let assign = CStmt::Assign {
        var,
        exp: CExp::Atom(atom(exp)),
      };
      CTail::Seq(assign, box cont)
    }
    Exp::Prim { op: (_, op), args } => {
      let prim = prim(op, args);
      let assign = CStmt::Assign {
        var,
        exp: CExp::Prim(prim),
      };
      CTail::Seq(assign, box cont)
    }
    Exp::Let { var: (_, var1), init: box (_, init1), body } => {
      explicate_assign(var1, init1, explicate_assign(var, body.1, cont))
    }
    exp => unimplemented!("unsupported form {:?}", exp)
  }
}

fn atom(exp: Exp<IdxVar>) -> CAtom {
  match exp {
    Exp::Int(n) => CAtom::Int(n),
    Exp::Var(var) => CAtom::Var(var),
    exp => unreachable!("{:?}", exp),
  }
}

fn prim(op: &str, mut args: Vec<(Range, Exp<IdxVar>)>) -> CPrim {
  match op {
    "read" => CPrim::Read,
    "-" => CPrim::Neg(atom(args.pop().unwrap().1)),
    "+" => CPrim::Add(
      atom(args[0].1.clone()),
      atom(args.pop().unwrap().1),
    ),
    _ => unreachable!("{}", op)
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use ast::*;
  use insta::assert_snapshot;

  #[test]
  fn let_in_init() {
    let prog = parse(r#"(let ([y (let ([x 20]) (+ x (let ([x 22]) x)))]) y)"#).unwrap();
    let prog = super::super::uniquify::uniquify(prog).unwrap();
    let prog = super::super::anf::anf(prog);
    let result = explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn nested_prims() {
    let prog = parse(r#"(let ([x (read)] [y (+ 2 3)]) (+ (- (read)) (+ y (- 2))))"#).unwrap();
    let prog = super::super::uniquify::uniquify(prog).unwrap();
    let prog = super::super::anf::anf(prog);
    let result = explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}