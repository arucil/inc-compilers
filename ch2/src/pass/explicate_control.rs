use asm::Label;
use ast::{Exp, ExpKind, IdxVar, Program};
use control::{CAtom, CExp, CPrim, CProgram, CStmt, CTail};
use indexmap::IndexSet;
use std::fmt::{self, Debug, Formatter};

pub struct CInfo {
  pub locals: IndexSet<IdxVar>,
}

impl Debug for CInfo {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    writeln!(f, "locals: {:?}\n", self.locals)
  }
}

pub fn explicate_control(mut prog: Program<IdxVar>) -> CProgram<CInfo> {
  let locals = collect_locals(&prog);
  CProgram {
    info: CInfo { locals },
    body: vec![(Label::Start, explicate_tail(prog.body.pop().unwrap()))],
  }
}

fn explicate_tail(exp: Exp<IdxVar>) -> CTail {
  match exp.kind {
    ExpKind::Int(_) | ExpKind::Var(_) => CTail::Return(CExp::Atom(atom(exp))),
    ExpKind::Prim { op: (_, op), args } => {
      CTail::Return(CExp::Prim(prim(op, args)))
    }
    ExpKind::Let { var, init, body } => {
      explicate_assign(var.1, *init, explicate_tail(*body))
    }
    exp => unimplemented!("unsupported form {:?}", exp),
  }
}

fn explicate_assign(var: IdxVar, init: ast::Exp<IdxVar>, cont: CTail) -> CTail {
  match init.kind {
    ExpKind::Int(_) | ExpKind::Var(_) => {
      let assign = CStmt::Assign {
        var,
        exp: CExp::Atom(atom(init)),
      };
      CTail::Seq(assign, box cont)
    }
    ExpKind::Prim { op: (_, op), args } => {
      let prim = prim(op, args);
      let assign = CStmt::Assign {
        var,
        exp: CExp::Prim(prim),
      };
      CTail::Seq(assign, box cont)
    }
    ExpKind::Let {
      var: (_, var1),
      init: box init1,
      body,
    } => explicate_assign(var1, init1, explicate_assign(var, *body, cont)),
    exp => unimplemented!("unsupported form {:?}", exp),
  }
}

fn atom(exp: Exp<IdxVar>) -> CAtom {
  match exp.kind {
    ExpKind::Int(n) => CAtom::Int(n),
    ExpKind::Var(var) => CAtom::Var(var),
    exp => unreachable!("{:?}", exp),
  }
}

fn prim(op: &str, mut args: Vec<Exp<IdxVar>>) -> CPrim {
  match op {
    "read" => CPrim::Read,
    "-" => {
      if args.len() == 1 {
        CPrim::Neg(atom(args.pop().unwrap()))
      } else {
        CPrim::Sub(atom(args[0].clone()), atom(args.pop().unwrap()))
      }
    }
    "+" => CPrim::Add(atom(args[0].clone()), atom(args.pop().unwrap())),
    _ => unreachable!("{}", op),
  }
}

fn collect_locals<TYPE>(prog: &Program<IdxVar, TYPE>) -> IndexSet<IdxVar> {
  let mut locals = IndexSet::new();
  for exp in &prog.body {
    collect_exp_locals(exp, &mut locals);
  }
  locals
}

fn collect_exp_locals<TYPE>(
  exp: &Exp<IdxVar, TYPE>,
  locals: &mut IndexSet<IdxVar>,
) {
  match &exp.kind {
    ExpKind::Int(_)
    | ExpKind::Var(_)
    | ExpKind::Str(_)
    | ExpKind::Bool(_)
    | ExpKind::Void
    | ExpKind::Get(_) => {}
    ExpKind::Let { var, init, body } => {
      locals.insert(var.1.clone());
      collect_exp_locals(&*init, locals);
      collect_exp_locals(&*body, locals);
    }
    ExpKind::Prim { op: _, args } => {
      for arg in args {
        collect_exp_locals(arg, locals);
      }
    }
    ExpKind::If { cond, conseq, alt } => {
      collect_exp_locals(&*cond, locals);
      collect_exp_locals(&*conseq, locals);
      collect_exp_locals(&*alt, locals);
    }
    ExpKind::Set { var: _, exp } => {
      collect_exp_locals(&*exp, locals);
    }
    ExpKind::Begin { seq, last } => {
      for exp in seq {
        collect_exp_locals(&*exp, locals);
      }
      collect_exp_locals(&*last, locals);
    }
    ExpKind::While { cond, body } => {
      collect_exp_locals(&*cond, locals);
      collect_exp_locals(&*body, locals);
    }
    ExpKind::Print(args) => {
      for arg in args {
        collect_exp_locals(arg, locals);
      }
    }
    ExpKind::NewLine => {}
  }
}

#[cfg(test)]
mod tests {
  use super::super::*;
  use super::*;
  use insta::assert_snapshot;

  #[test]
  fn let_in_init() {
    let prog = ast::parse(
      r#"
        (let
          ([y (let ([x 20])
                (+ x
                   (let ([x 22]) x)))])
          y)
        "#,
    )
    .unwrap();
    let prog = uniquify::uniquify(prog).unwrap();
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn nested_prims() {
    let prog = ast::parse(
      r#"(let ([x (read)] [y (+ 2 3)]) (+ (- (read)) (+ y (- 2))))"#,
    )
    .unwrap();
    let prog = uniquify::uniquify(prog).unwrap();
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
