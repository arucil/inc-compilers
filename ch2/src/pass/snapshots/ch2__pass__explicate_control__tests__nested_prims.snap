---
source: ch2/src/pass/explicate_control.rs
expression: result.to_string_pretty()
---
locals: {x.0, y.1, (tmp).0, (tmp).1, (tmp).2, (tmp).3}

.start:
    x.0 = (read)
    y.1 = (+ 2 3)
    (tmp).0 = (read)
    (tmp).1 = (- (tmp).0)
    (tmp).2 = (- 2)
    (tmp).3 = (+ y.1 (tmp).2)
    return (+ (tmp).1 (tmp).3)

